### PR TITLE
Fix encoding bug in `gensim.scripts.word2vec2tensor`. Fix #1958

### LINKED ISSUE #1958: gensim.scripts.word2vec2tensor TypeError: write() argument must be str, not bytes
**How I make article_body_w2v_300.txt**  

```
import gensim
from gensim.models import Word2Vec
from gensim.models.word2vec import LineSentence

sentences = LineSentence("./data/article_body_corpus.txt")

model = Word2Vec(sentences, size=300, window=5, min_count=5, workers=4)

model.wv.save_word2vec_format("article_body_w2v_300.txt", binary=False)
```

**Command I use to run gensim.scripts.word2vec2tensor ** 
python -m gensim.scripts.word2vec2tensor -i article_body_w2v_300.txt -o meow/
2018-03-07 16:30:29,484 - word2vec2tensor - INFO - running 

Console output
```
word_embedding  python -m gensim.scripts.word2vec2tensor -i article_body_w2v_300.txt -o meow/
2018-03-07 16:30:29,484 - word2vec2tensor - INFO - running /home/cpu11453local/anaconda3/envs/gensim/lib/python3.6/site-packages/gensim/scripts/word2vec2tensor.py -i article_body_w2v_300.txt -o meow/
2018-03-07 16:30:29,484 - utils_any2vec - INFO - loading projection weights from article_body_w2v_300.txt
2018-03-07 16:30:41,992 - utils_any2vec - INFO - loaded (56543, 300) matrix from article_body_w2v_300.txt
Traceback (most recent call last):
  File "/home/cpu11453local/anaconda3/envs/gensim/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/cpu11453local/anaconda3/envs/gensim/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/cpu11453local/anaconda3/envs/gensim/lib/python3.6/site-packages/gensim/scripts/word2vec2tensor.py", line 93, in <module>
    word2vec2tensor(args.input, args.output, args.binary)
  File "/home/cpu11453local/anaconda3/envs/gensim/lib/python3.6/site-packages/gensim/scripts/word2vec2tensor.py", line 73, in word2vec2tensor
    file_metadata.write(gensim.utils.to_utf8(word) + gensim.utils.to_utf8('\n'))
TypeError: write() argument must be str, not bytes

```

### PR DESCRIPTION
**Pull request by AakaashRao**
This PR fixes #1958, which noted an issue with the Python 3 (but not Python 2) version of the `word2vec2tensor` function. The problematic line was attempting to encode a string in `utf-8` and write the resulting bytestring to file, causing an error; this revised code simply writes the original string. I have confirmed that the code runs as expected on Python 2.7 and Python 3.6.

CC @menshikh-iv.

### CODE REVIEW

**Comment by menshikh-iv** [comment]
Hello @AakaashRao 

1. It's better to write all as bytes always (this is some our convention)
2. `.DS_store` is unrelated file, please remove it
3. Definitely, need to add a tests for this functionality 
4. Replace `open` to `smart_open`

File: gensim/scripts/word2vec2tensor.py [hunk 1]
```diff
@@ -63,22 +64,25 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
         True if input file in binary format.
 
     """
+
```
Comment (menshikh-iv) [inline gensim/scripts/word2vec2tensor.py]: style-nitpick: no need an empty line here

File: gensim/scripts/word2vec2tensor.py [hunk 2]
```diff
@@ -63,22 +64,25 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
...
     outfiletsv = tensor_filename + '_tensor.tsv'
     outfiletsvmeta = tensor_filename + '_metadata.tsv'
 
-    with open(outfiletsv, 'w+') as file_vector:
-        with open(outfiletsvmeta, 'w+') as file_metadata:
+    with utils.smart_open(outfiletsv, 'wb') as file_vector:
+        with utils.smart_open(outfiletsvmeta, 'wb') as file_metadata:
             for word in model.index2word:
-                file_metadata.write(gensim.utils.to_utf8(word) + gensim.utils.to_utf8('\n'))
+                write = gensim.utils.any2utf8(word) + gensim.utils.any2utf8('\n')
```
Comment (menshikh-iv) [inline gensim/scripts/word2vec2tensor.py]: no needed changes (this is almost same), you can return an old variant (except for last line in cycle)

File: gensim/test/test_word2vec2tensor.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import numpy
+import gensim
+
+from gensim.utils import check_output, smart_open
+from gensim.test.utils import get_tmpfile
+
+
+class TestWord2Vec2Tensor(unittest.TestCase):
+    def setUp(self):
+        # self.datapath = datapath('test_word2vec.txt')
```
Comment (menshikh-iv) [inline gensim/test/test_word2vec2tensor.py]: this is right way how to retrieve path

File: gensim/test/test_word2vec2tensor.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        self.datapath = 'test/test_data/test_word2vec.txt'  # temporary
+        self.output_folder = get_tmpfile('')
+        self.metadata_file = self.output_folder + '_metadata.tsv'
+        self.tensor_file = self.output_folder + '_tensor.tsv'
+        self.vector_file = self.output_folder + '_vector.tsv'
+
+    def testConversion(self):
+        check_output(args=[
+            sys.executable, '-m', 'scripts.word2vec2tensor', '--input',
```
Comment (menshikh-iv) [inline gensim/test/test_word2vec2tensor.py]: you should test changed function `word2vec2tensor`, not all script

File: gensim/test/test_word2vec2tensor.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            with smart_open(self.metadata_file, 'rb') as f:
+                self.metadata = f.readlines()
+        except FileNotFoundError:
+            self.fail(
+                'Metadata file %s creation failed. \
+                Check the parameters and input file format.' % self.metadata_file
+                )
+        try:
+            with smart_open(self.tensor_file, 'rb') as f:
+                self.vectors = f.readlines()
```
Comment (menshikh-iv) [inline gensim/test/test_word2vec2tensor.py]: I don't catch, why you need `self` here (same question for `number_words`, `metadata`, etc)?

File: gensim/test/test_word2vec2tensor.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        self.vectors = [vector.replace(b'\t', b' ') for vector in self.vectors]
+        self.word2veclines = [self.metadata[i] + b' ' + self.vectors[i] for i in range(len(self.metadata))]
+        with smart_open(self.vector_file, 'wb') as f:
+            # write header
+            f.write(gensim.utils.any2utf8(str(self.number_words) + ' ' + str(self.vector_size) + '\n'))
+            f.writelines(self.word2veclines)
+
+        # test that the converted model loads successfully
+        self.test_model = gensim.models.KeyedVectors.load_word2vec_format(self.vector_file, binary=False)
+        self.assertTrue(numpy.allclose(self.test_model.n_similarity(['the', 'and'], ['and', 'the']), 1.0))
```
Comment (menshikh-iv) [inline gensim/test/test_word2vec2tensor.py]: +1 for this test, but -1 for this type of assertion.
If you want to check that models are equivalent - you should be sure, that vectors for each word are equivalent.

File: gensim/test/test_word2vec2tensor.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+if __name__ == '__main__':
+    logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.DEBUG)
```
Comment (menshikh-iv) [inline gensim/test/test_word2vec2tensor.py]: please use `logging.INFO` instead of `DEBUG`

**Comment by AakaashRao** [comment]
Thanks @menshikh-iv for the feedback. I've addressed your points and written some tests. A couple of quick points/questions about the tests:

1. I've added some test word2vec vectors into `test_data` as `test_data/test_word2vec.txt` because I couldn't find any vectors saved already. Let me know if this is incorrect.
2. Even in a `virtualenv`, I could not get the `datapath` function -- which is supposed to prepend the `test_data` directory to a filename -- to look in the `test_data` folder of my development version of gensim (containing the new `test_word2vec.txt` file) rather than in the `test_data` in my installed gensim. For now I have replaced it with a temporary fix, but I would appreciate any guidance -- I'm sure I'm missing something obvious.

**Comment by menshikh-iv** [comment]
@AakaashRao 
1. For example https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_data/word2vec_pre_kv_c (large) or https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_data/high_precision.kv.txt (small)
2. Yes, you have a very old fork, for fixing this, you should to fetch fresh upstream and merge `upstream/develop` to your current branch.

you stay in PR branch
```bash
git remote add upstream https://github.com/RaRe-Technologies/gensim.git
git fetch upstream
git merge upstream/develop
```

**Comment by AakaashRao** [comment]
Thanks @menshikh-iv. I've addressed your review. Like before, I had some problems getting Python to import the current working version of `word2vec2tensor.py` instead of the version I had installed from the master version of Gensim (even when using the virtualenv). As a temporary fix, I placed a `scripts` folder containing `word2vec2tensor.py` in `test`. I would really appreciate if you could tell me how to fix this issue.

File: gensim/scripts/word2vec2tensor.py [hunk 8]
```diff
@@ -67,12 +68,12 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
     outfiletsv = tensor_filename + '_tensor.tsv'
     outfiletsvmeta = tensor_filename + '_metadata.tsv'
 
-    with open(outfiletsv, 'w+') as file_vector:
-        with open(outfiletsvmeta, 'w+') as file_metadata:
+    with utils.smart_open(outfiletsv, 'wb') as file_vector:
+        with utils.smart_open(outfiletsvmeta, 'wb') as file_metadata:
             for word in model.index2word:
-                file_metadata.write(gensim.utils.to_utf8(word) + gensim.utils.to_utf8('\n'))
+                file_metadata.write(gensim.utils.any2utf8(word) + gensim.utils.any2utf8('\n'))
```
Comment (menshikh-iv) [inline gensim/scripts/word2vec2tensor.py]: why `any2utf8` instead of `to_utf8`?

File: gensim/test/test_scripts.py [hunk 9]
```diff
@@ -103,6 +107,75 @@ def test_segment_and_write_all_articles(self):
...
+            self.fail(
+                'Metadata file %s creation failed. \
+                Check the parameters and input file format.' % self.metadata_file
+                )
+        try:
+            with smart_open(self.tensor_file, 'rb') as f:
+                vectors = f.readlines()
+        except FileNotFoundError:
+            self.fail(
+                'Tensor file %s creation failed. \
```
Comment (menshikh-iv) [inline gensim/test/test_scripts.py]: carriage-return  (here and everywhere)

File: gensim/test/scripts/word2vec2tensor.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
```
Comment (menshikh-iv) [inline gensim/test/scripts/word2vec2tensor.py]: `from gensim.scripts.word2vec2tensor import word2vec2tensor` should work fine (without this file).

**Comment by menshikh-iv** [comment]
Looks abandoned, I close this.
@AakaashRao feel free to re-open when you be ready to finish it.

**Comment by martinnormark** [comment]
Oh darn, this would've been useful to get merged.

Is it only the last three review comments, that prevents this from getting merged?

FWIW, I manually fixed this in a separate conda environment by adding `decode` calls:
```python
file_metadata.write(gensim.utils.to_utf8(word).decode("utf-8") + gensim.utils.to_utf8('\n').decode("utf-8"))
```

**Comment by menshikh-iv** [comment]
@martinnormark mostly yes, feel free to submit new PR :+1:

**Pull request closed without merging**